File: packages/ai-config/src/secrets/resolver.ts
import type { AuthSource, SecretRef } from '../domain/auth.js';

export interface SecretResolverContext {
  env?: Record<string, string | undefined>;
  profiles?: Record<string, Record<string, string>>;
  profileName?: string;
}

function resolveRef(ref: SecretRef | string, ctx: SecretResolverContext): string | undefined {
  if (typeof ref === 'string') {
    return (ctx.env ?? process.env)[ref];
  }

  if (ref.type === 'env' && ref.name) {
    return (ctx.env ?? process.env)[ref.name];
  }

  if (ref.type === 'file') {
    return undefined;
  }

  if (ref.type === 'exec') {
    return undefined;
  }

  return undefined;
}

export function resolveAuthSecret(
  auth: AuthSource,
  providerId: string,
  context: SecretResolverContext = {},
): AuthSource {
  const env = context.env ?? process.env;
  const profileSecrets = context.profileName ? (context.profiles?.[context.profileName] ?? {}) : {};
  const normalizedProviderId = providerId.toUpperCase().replace(/[^A-Z0-9]/g, '_');
  const envKey = `${normalizedProviderId}_API_KEY`;

  const envValue = env[envKey];
  if (envValue) {
    return { ...auth, mode: 'api_key', value: envValue, from: 'env', redacted: true };
  }

  const profileValue = profileSecrets[providerId] ?? profileSecrets[envKey];
  if (profileValue) {
    return { ...auth, mode: 'api_key', value: profileValue, from: 'profile', redacted: true };
  }

  if (auth.ref) {
    const refValue = resolveRef(auth.ref, context);
    if (refValue) {
      return { ...auth, mode: 'api_key', value: refValue, from: 'ref', redacted: true };
    }
  }

  if (auth.value || auth.token) {
    return { ...auth, from: 'config', redacted: true };
  }

  return { ...auth, from: 'config', redacted: false };
}

File: packages/ai_config/secrets/resolver.py
import os
from typing import Dict, Optional

from ..domain.auth import AuthSource


def _resolve_ref(ref, env: Dict[str, str]) -> Optional[str]:
    if isinstance(ref, str):
        return env.get(ref)

    ref_type = ref.get("type")
    if ref_type == "env":
        return env.get(ref.get("name", ""))

    if ref_type in ["file", "exec"]:
        return None

    return None


def resolve_auth_secret(
    auth: AuthSource,
    provider_id: str,
    env: Optional[Dict[str, str]] = None,
    profiles: Optional[Dict[str, Dict[str, str]]] = None,
    profile_name: Optional[str] = None,
) -> AuthSource:
    merged_env = dict(os.environ)
    merged_env.update(env or {})

    normalized_provider_id = "".join(c if c.isalnum() else "_" for c in provider_id.upper())
    env_key = f"{normalized_provider_id}_API_KEY"

    env_value = merged_env.get(env_key)
    if env_value:
        return AuthSource(mode="api_key", value=env_value, ref=auth.ref)

    profile_bucket = (profiles or {}).get(profile_name or "", {})
    profile_value = profile_bucket.get(provider_id) or profile_bucket.get(env_key)
    if profile_value:
        return AuthSource(mode="api_key", value=profile_value, ref=auth.ref)

    if auth.ref:
        ref_value = _resolve_ref(auth.ref, merged_env)
        if ref_value:
            return AuthSource(mode="api_key", value=ref_value, ref=auth.ref)

    return auth
